### PR TITLE
datetime: make date.new{} and date:set{} equivalent

### LINKED ISSUE #6793: Datetime module quirks
Passing the same parameters when creating/modifying a datatime object should give the same result no matter how those parameters were passed. It is not the case now:

```lua
unix/:/var/run/tarantool/tarantool.sock> box.info.version
---
- 2.10.0-beta2-0-g9d80b68b9
...

unix/:/var/run/tarantool/tarantool.sock> date = require('datetime')
---
...

unix/:/var/run/tarantool/tarantool.sock> date.new({ tzoffset = '+0800', timestamp = 1630359071 })
---
- 2021-08-30T21:31:11+0800
...

unix/:/var/run/tarantool/tarantool.sock> date.new():set({ tzoffset = '+0800', timestamp = 1630359071 })
---
- 2021-08-30T21:31:11Z
...

unix/:/var/run/tarantool/tarantool.sock> date.new({tzoffset = '+0800'}):set({ tzoffset = '+0800', timestamp = 1630359071 })
---
- 2021-08-31T05:31:11+0800
...
```

**Actual behavior**

3 different dates are printed.

**Expected behavior**

The same date is printed for all three datetime objects.

@tsafin confirmed that this is a bug:

> me> так и должно быть?
> tsafin> не, неправильно. Это ordering в обработчке timestamp и tzoffset (мы обработав timestamp сваливаем раньше, чем успеваем обработать tzoffset)

### PR DESCRIPTION
**Pull request by tsafin**
Constructor date.new() and modifier date:set() should always produce same
result for all attributes combinations. Fixed the problem for
`timestamp` with `tzoffset`.

```
tarantool> date.new{ tzoffset = '+0800', timestamp = 1630359071 }
---
- 2021-08-30T21:31:11+0800
...

tarantool> date.new():set{ tzoffset = '+0800', timestamp = 1630359071 }
---
- 2021-08-30T21:31:11+0800
...

```

> Discovered that there was a similar problem with `:set{}` using `tzoffset` attribute not only 
> for `timestamp` but with all the rest date or time attributes. So have modified code flow to
> always "delocalize"/localize time zone upon any relevant attribute manipulation.

Fixes #6793

### CODE REVIEW

**Review by ligurio — CHANGES_REQUESTED** [review]

File: changelogs/unreleased/gh-6821-datetime-set-with-tzoffset.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+## bugfix/datetime
+
+ * Fixed a bug in datetime module when `date:set{tzoffset=XXX}` was not
```
[the remark this answers is not in the document]
Comment (tsafin) [reply changelogs/unreleased/gh-6821-datetime-set-with-tzoffset.md]: Fixed

File: test/app-tap/datetime.test.lua [hunk 2]
```diff
@@ -1163,20 +1163,52 @@ test:test("Time :set{} operations", function(test)
...
+    test:plan(9)
+    local ts, ts2
+    local obj = {}
+    local attribs = {
+        {'year', 2000},
+        {'month', 11},
+        {'day', 30},
+        {'hour', 6},
+        {'min', 12},
+        {'sec', 23},
```
[the remark this answers is not in the document]
Comment (tsafin) [reply test/app-tap/datetime.test.lua:1695]: Probably only `nsec` (because every next iteration this tests remember new attribute to check, and if `usec`/`msec` would be defined at the same time as `nsec` then we would get error. And I see no much point in overcomplicating this test to handle this corner case with `nsec`/`usec`/`msec`)

**Review by ligurio — CHANGES_REQUESTED** [review]

File: changelogs/unreleased/gh-6821-datetime-set-with-tzoffset.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+   producing the same result with `date.new{tzoffset=XXX}` for the same
+   set of attributes passed (gh-6821);
```
[the remark this answers is not in the document]
Comment (tsafin) [reply changelogs/unreleased/gh-6821-datetime-set-with-tzoffset.md:5]: Good point - fixed. Thanks!

[the remark this answers is not in the document]
Comment (tsafin) [reply changelogs/unreleased/gh-6821-datetime-set-with-tzoffset.md:5]: Corrected

**Comment by tsafin** [comment]
There was a merge conflict (after recent datetime comit for 'tz' attribute) which required some attention, so I've resolved conflict and added single case of checking 'tz' attribute in a test named "Check :set{} and .new{} equal for all attributes". Now it's feature complete. @ligurio

File: test/app-tap/datetime.test.lua [hunk 4]
```diff
@@ -1667,20 +1667,54 @@ test:test("Time :set{} operations", function(test)
...
             'msec = 123')
-    test:is(tostring(ts:set{ usec = 123}), '2021-08-31T05:31:11.000123+0800',
+    test:is(tostring(ts:set{ usec = 123}), '2021-08-30T21:31:11.000123+0800',
             'usec = 123')
-    test:is(tostring(ts:set{ nsec = 123}), '2021-08-31T05:31:11.000000123+0800',
+    test:is(tostring(ts:set{ nsec = 123}), '2021-08-30T21:31:11.000000123+0800',
             'nsec = 123')
 end)
 
+test:test("Check :set{} and .new{} equal for all attributes", function(test)
```
Comment (ligurio) [inline test/app-tap/datetime.test.lua:1685]: `.new` -> `:new`

**Review by ligurio — APPROVED** [review]
LGTM after fixing typo in testcase name (se comment)

File: test/app-tap/datetime.test.lua (hunk 4 above)
Comment (tsafin) [reply test/app-tap/datetime.test.lua:1685]: No, `date.new` could not be an instance method, that's why this syntactic difference to `dateobj:set{}` (which has to be called on object instance).

Comment (ligurio) [reply test/app-tap/datetime.test.lua:1685]: Ah, right.
